components/launcher/src/server/mod.rs

```diff
@@ -214,7 +214,31 @@ impl Server {
                 );
             }
         }
-        self.supervisor.wait().ok();
+
+        // With the Supervisor shutting down services, we need to
+        // ensure that the Launcher is able to reap those main service
+        // processes as the Supervisor shuts them down... otherwise,
+        // the Supervisor can end up waiting a long time on zombie
+        // processes.
+        //
+        // But see https://github.com/habitat-sh/habitat/issues/6131
+        // for a possible future where this isn't needed, and reaping
+        // could theoretically just take place at the very end of this
+        // shutdown process, rather than repeatedly.
+        while let Ok(None) = self.supervisor.try_wait() {
+            self.services.reap_services();
+            thread::sleep(Duration::from_millis(5));
+        }
+
+        match self.supervisor.try_wait() {
+            Ok(Some(status)) => debug!("Supervisor exited with: {}", status),
+            Err(e) => error!("Error waiting on supervisor: {:?}", e),
+            __ => unreachable!(),
+        }
+
+        // TODO (CM): Eventually this can go away... but we need to
+        // keep it around while we still support older Supervisors
+        // that don't shutdown services themselves.
```

Comment:
It should also be noted that this needs to stay until we get all the other hooks (especially `init`) running in their own thread to support windows supervisors with `IGNORE_SIGNALS` enabled (which is the default when running in a windows service). With `IGNORE_SIGNALS` the supervisor is forcefully terminated with a `ctrl+c` event and thus the launcher needs to terminate the processes. We enable `IGNORE_SIGNALS` to accomodate long running `init` hooks which block the sup service and its ability to respond to signals.